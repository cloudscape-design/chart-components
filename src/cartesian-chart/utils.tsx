// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import type Highcharts from "highcharts";

import { warnOnce } from "@cloudscape-design/component-toolkit/internal";

import { CartesianChartProps, InternalSeriesOptions, InternalTooltipMatchedItem } from "./interfaces-cartesian";

export function getDataExtremes(axis?: Highcharts.Axis): [number, number] {
  const extremes = axis?.getExtremes();
  return [extremes?.dataMin ?? 0, extremes?.dataMax ?? 0];
}

export function findMatchedTooltipItems(point: Highcharts.Point, series: InternalSeriesOptions[]) {
  const matchedItems: InternalTooltipMatchedItem[] = [];
  function findMatchedItem(series: InternalSeriesOptions) {
    if (series.type === "x-threshold" || series.type === "y-threshold") {
      return;
    }
    if ("data" in series && series.data && Array.isArray(series.data)) {
      for (let i = 0; i < series.data.length; i++) {
        const detail = getMatchedTooltipItemForIndex(i, series);
        if (detail?.x === point.x) {
          matchedItems.push(detail);
        }
      }
    }
  }
  series.forEach((s) => findMatchedItem(s));

  function findMatchedXThreshold(series: CartesianChartProps.XThresholdSeriesOptions) {
    if (series.value <= point.x && point.x <= series.value) {
      matchedItems.push({ type: "all", x: series.value, series });
    }
  }
  series.forEach((s) => (s.type === "x-threshold" ? findMatchedXThreshold(s) : undefined));

  function findMatchedYThreshold(series: CartesianChartProps.YThresholdSeriesOptions) {
    matchedItems.push({ type: "point", x: point.x, y: series.value, series });
  }
  series.forEach((s) => (s.type === "y-threshold" ? findMatchedYThreshold(s) : undefined));

  return matchedItems;
}

function getMatchedTooltipItemForIndex(
  index: number,
  series: InternalSeriesOptions,
): null | InternalTooltipMatchedItem {
  const x = getSeriesXbyIndex(series, index);
  const y = getSeriesYbyIndex(series, index);
  if (x === null) {
    return null;
  }
  if (typeof y === "number") {
    return { type: "point", x: x, y, series };
  }
  if (Array.isArray(y)) {
    return { type: "range", x: x, low: y[0], high: y[1], series };
  }
  if (series.type === "y-threshold") {
    return { type: "all", x: x, series };
  }
  return null;
}

function getSeriesXbyIndex(series: InternalSeriesOptions, index: number): number | null {
  if (!("data" in series) || !Array.isArray(series.data)) {
    warnOnce("CartesianChart", "Series data cannot be parsed.");
    return null;
  }
  switch (series.type) {
    case "area":
    case "areaspline":
    case "column":
    case "line":
    case "scatter":
    case "spline": {
      const item = series.data[index];
      if (Array.isArray(item) && typeof item[0] === "number") {
        return item[0];
      }
      if (item && typeof item === "object" && !Array.isArray(item) && typeof item.x === "number") {
        return item.x;
      }
      return index;
    }
    case "errorbar": {
      const item = series.data[index];
      if (Array.isArray(item) && typeof item[0] === "number") {
        return item.length === 3 ? item[0] : index;
      }
      if (item && typeof item === "object" && !Array.isArray(item) && typeof item.x === "number") {
        return item.x ?? index;
      }
      return index;
    }
    default:
      return index;
  }
}

function getSeriesYbyIndex(series: InternalSeriesOptions, index: number): null | number | [number, number] {
  if (!("data" in series) || !Array.isArray(series.data)) {
    warnOnce("CartesianChart", "Series data cannot be parsed.");
    return null;
  }
  switch (series.type) {
    case "area":
    case "areaspline":
    case "column":
    case "line":
    case "scatter":
    case "spline": {
      const item = series.data[index];
      if (Array.isArray(item)) {
        return item[1];
      }
      if (typeof item === "number") {
        return item;
      }
      if (item && typeof item === "object") {
        return item.y ?? null;
      }
      return null;
    }
    case "errorbar": {
      const item = series.data[index];
      if (Array.isArray(item)) {
        const [low, high] = item.length === 2 ? [item[0], item[1]] : [item[1], item[2]];
        return typeof low === "number" ? [low, high] : null;
      }
      if (item && typeof item === "object") {
        const [low, high] = [item.low, item.high];
        return typeof low === "number" && typeof high === "number" ? [low, high] : null;
      }
      return null;
    }
    default:
      return null;
  }
}
