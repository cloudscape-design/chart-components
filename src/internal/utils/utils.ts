// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export type SomeRequired<Type, Keys extends keyof Type> = Type & {
  [Key in Keys]-?: Type[Key];
};

// Highcharts series and data arrays are not marked as readonly in TS, but are readonly effectively.
// This creates a conflict with Cloudscape type definitions as we use readonly arrays on input.
// To resolve the conflict, we cast out readonly types to writeable.
export type Writeable<T> = { -readonly [P in keyof T]: T[P] };

export function castArray<T>(valueOrArray?: T | T[]): undefined | T[] {
  if (!valueOrArray) {
    return undefined;
  }
  if (Array.isArray(valueOrArray)) {
    return valueOrArray;
  }
  return [valueOrArray];
}

export function isEqualArrays<T>(a: readonly T[], b: readonly T[], eq: (a: T, b: T) => boolean) {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (!eq(a[i], b[i])) {
      return false;
    }
  }
  return true;
}

export class DebouncedCall {
  private locked = false;
  private lockTimeoutRef = setTimeout(() => {}, 0);
  private timeoutRef = setTimeout(() => {}, 0);

  public call(callback: () => void, delay = -1) {
    if (!this.locked) {
      this.cancelPrevious();
      if (delay === -1) {
        callback();
      } else {
        this.timeoutRef = setTimeout(callback, delay);
      }
    }
  }

  public cancelPrevious() {
    clearTimeout(this.timeoutRef);
  }

  public lock(timeout: number) {
    clearTimeout(this.lockTimeoutRef);
    this.locked = true;
    this.lockTimeoutRef = setTimeout(() => (this.locked = false), timeout);
  }
}
