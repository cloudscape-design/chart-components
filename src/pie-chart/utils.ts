// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { getOptionsId } from "../core/utils";
import { InternalSeriesOptions } from "./interfaces-pie";

export function getAllSegmentIds(series: InternalSeriesOptions[]): string[] {
  const allSegmentIds = series.flatMap((s) => {
    const itemIds: string[] = [];
    if ((s.type === "pie" || s.type === "donut") && s.data) {
      for (const dataItem of s.data) {
        if (dataItem && typeof dataItem === "object" && !Array.isArray(dataItem)) {
          itemIds.push(getOptionsId(dataItem));
        }
      }
    }
    return itemIds;
  });
  return allSegmentIds;
}
