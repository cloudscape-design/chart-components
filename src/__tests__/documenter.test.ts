// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { expect, test } from "vitest";

import componentDefinitions from "../../lib/components/internal/api-docs/components";
import { getAllComponents } from "./utils";

test.each<string>(getAllComponents())(`definition for %s matches the snapshot`, (componentName: string) => {
  const definition = componentDefinitions[componentName];
  expect(definition).toMatchSnapshot(componentName);
});
