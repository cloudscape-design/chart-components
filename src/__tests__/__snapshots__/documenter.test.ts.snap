// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`definition for cartesian-chart matches the snapshot > cartesian-chart 1`] = `
{
  "dashCaseName": "cartesian-chart",
  "events": [
    {
      "cancelable": false,
      "description": "A callback function, triggered when series visibility changes as a result of user interaction with the legend or filter.",
      "detailInlineType": {
        "name": "{ visibleSeries: Array<string>; }",
        "properties": [
          {
            "name": "visibleSeries",
            "optional": false,
            "type": "Array<string>",
          },
        ],
        "type": "object",
      },
      "detailType": "{ visibleSeries: Array<string>; }",
      "name": "onVisibleSeriesChange",
    },
  ],
  "functions": [
    {
      "description": "Controls series visibility and works with both controlled and uncontrolled visibility modes.",
      "name": "setVisibleSeries",
      "parameters": [
        {
          "name": "visibleSeries",
          "type": "ReadonlyArray<string>",
        },
      ],
      "returnType": "void",
    },
    {
      "description": "Functions similarly to \`setVisibleSeries\`, but applies to all series and doesn't require series IDs as input.
Use this when implementing clear-filter actions in no-match states.",
      "name": "showAllSeries",
      "parameters": [],
      "returnType": "void",
    },
  ],
  "name": "CartesianChart",
  "properties": [
    {
      "description": "Defines the ARIA description of the chart container.
This property corresponds to [accessibility.description](https://api.highcharts.com/highcharts/accessibility.description),
and requires the [accessibility module](https://www.highcharts.com/docs/accessibility/accessibility-module).",
      "name": "ariaDescription",
      "optional": true,
      "type": "string",
    },
    {
      "description": "Defines the ARIA label for the chart container.
This property corresponds to [lang.chartContainerLabel](https://api.highcharts.com/highcharts/lang.accessibility.chartContainerLabel),
and requires the [accessibility module](https://www.highcharts.com/docs/accessibility/accessibility-module).",
      "name": "ariaLabel",
      "optional": true,
      "type": "string",
    },
    {
      "description": "The height of the chart plot in pixels. It does not include legend and filter.",
      "name": "chartHeight",
      "optional": true,
      "type": "number",
    },
    {
      "description": "Defines the minimum allowed height of the chart plot. Use this when \`fitHeight=true\` to
prevent the chart plot from becoming too small to display its content. When the parent
container is smaller than the minimum height, a vertical scrollbar appears automatically.",
      "name": "chartMinHeight",
      "optional": true,
      "type": "number",
    },
    {
      "description": "Defines the minimum allowed width of the chart plot. When the parent container is smaller the
minimum width, the horizontal scrollbar is automatically added.",
      "name": "chartMinWidth",
      "optional": true,
      "type": "number",
    },
    {
      "defaultValue": "true",
      "description": "When set to \`true\`, adds a visual emphasis on the zero baseline axis.",
      "name": "emphasizeBaseline",
      "optional": true,
      "type": "boolean",
    },
    {
      "description": "Defines options for filtering in the chart, including:
* \`seriesFilter\` (optional, boolean) - Displays default series filter at the top of the chart.
* \`additionalFilters\` (optional, slot) - A slot for custom chart filters at the top of the chart.",
      "inlineType": {
        "name": "BaseFilterOptions",
        "properties": [
          {
            "name": "additionalFilters",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "seriesFilter",
            "optional": true,
            "type": "boolean",
          },
        ],
        "type": "object",
      },
      "name": "filter",
      "optional": true,
      "type": "BaseFilterOptions",
    },
    {
      "defaultValue": "false",
      "description": "The chart automatically adjusts its height to fill the parent container when this property is set.",
      "name": "fitHeight",
      "optional": true,
      "type": "boolean",
    },
    {
      "description": "The Highcharts instance, which can be obtained using \`import Highcharts from "highcharts"\`.
Supported Highcharts versions: 12.",
      "name": "highcharts",
      "optional": false,
      "type": "object | null",
    },
    {
      "description": "An object that contains all of the localized strings required by the component.",
      "i18nTag": true,
      "inlineType": {
        "name": "CartesianI18nStrings",
        "properties": [
          {
            "name": "chartAccessibleDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "detailPopoverDismissAriaLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "errorText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "legendAriaLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "loadingText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "recoveryText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterPlaceholder",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterSelectedAriaLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "xAxisAccessibleDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "yAxisAccessibleDescription",
            "optional": true,
            "type": "string",
          },
        ],
        "type": "object",
      },
      "name": "i18nStrings",
      "optional": true,
      "type": "CartesianI18nStrings",
    },
    {
      "defaultValue": "false",
      "description": "Inverts X and Y axes. Use it to show horizontal columns (bars).
This property corresponds to [chart.inverted](https://api.highcharts.com/highcharts/chart.inverted).",
      "name": "inverted",
      "optional": true,
      "type": "boolean",
    },
    {
      "description": "Defines chart legend options, including:
* \`enabled\` (optional, boolean) - Hides legend when set to \`false\`.
* \`title\` (optional, string) - Visible label, shown above the legend.
* \`actions\` (optional, slot) - A slot before the legend that can be used to render custom actions.",
      "inlineType": {
        "name": "BaseLegendOptions",
        "properties": [
          {
            "name": "actions",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "enabled",
            "optional": true,
            "type": "boolean",
          },
          {
            "name": "title",
            "optional": true,
            "type": "string",
          },
        ],
        "type": "object",
      },
      "name": "legend",
      "optional": true,
      "type": "BaseLegendOptions",
    },
    {
      "description": "Defines options to represent empty, no-match, loading, and error state of the chart, including:
* \`statusType\` (optional, "finished" | "loading" | "error") - Specifies the current status of loading data.
* \`empty\` (optional, ReactNode) - Content displayed when the chart data is empty.
* \`noMatch\` (optional, ReactNode) - Content displayed when there is no data to display due to the built-in filtering.
* \`loading\` (optional, ReactNode) - Content displayed when \`statusType="loading"\`. If omitted, the default loading state
is shown, using \`i18n.loadingText\` or built-in i18n.
* \`error\` (optional, ReactNode) - Content displayed when \`statusType="error"\`. If omitted, the default error state
is shown, using \`i18n.errorText\` and \`i18n.recoveryText\` (when \`onRecoveryClick\` is provided), or built-in i18n.
* \`onRecoveryClick\` (optional, function) - Called when the user clicks the recovery button that appears when using default error
state, and only if \`onRecoveryClick\` is provided. Use this to enable the user to retry a failed request or provide another option
for the user to recover from the error.",
      "inlineType": {
        "name": "BaseNoDataOptions",
        "properties": [
          {
            "name": "empty",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "error",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "loading",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "noMatch",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "onRecoveryClick",
            "optional": true,
            "type": "NonCancelableEventHandler<{}>",
          },
          {
            "name": "statusType",
            "optional": true,
            "type": ""finished" | "loading" | "error"",
          },
        ],
        "type": "object",
      },
      "name": "noData",
      "optional": true,
      "type": "BaseNoDataOptions",
    },
    {
      "description": "Defines series options of the chart.
This property corresponds to [series](https://api.highcharts.com/highcharts/series), and extends it
with two additional series types: "x-threshold", and "y-threshold".

Supported types:
* [area](https://api.highcharts.com/highcharts/series.area).
* [areaspline](https://api.highcharts.com/highcharts/series.areaspline).
* [column](https://api.highcharts.com/highcharts/series.column).
* [errorbar](https://api.highcharts.com/highcharts/series.errorbar) - requires "highcharts/highcharts-more" module.
* [line](https://api.highcharts.com/highcharts/series.line).
* [scatter](https://api.highcharts.com/highcharts/series.scatter).
* [spline](https://api.highcharts.com/highcharts/series.spline).
* x-threshold - The line-like series to represent x-axis threshold (vertical, when \`inverted=false\`).
* y-threshold - The line-like series to represent y-axis threshold (horizontal, when \`inverted=false\`).",
      "name": "series",
      "optional": false,
      "type": "ReadonlyArray<CartesianChartProps.SeriesOptions>",
    },
    {
      "defaultValue": "undefined",
      "description": "Enables series stacking behavior. Use it for column- or area- series.
This property corresponds to "normal" stacking type in Highcharts ([plotOptions.series.stacking](https://api.highcharts.com/highcharts/plotOptions.series.stacking)).",
      "name": "stacking",
      "optional": true,
      "type": ""normal"",
    },
    {
      "description": "Defines tooltip options of the chart, including:
* \`enabled\` - (optional, boolean) - Hides the tooltip.
* \`size\` - (optional, "small" | "medium" | "large") - Specifies max tooltip size.
* \`placement\` - (optional, "middle" | "outside") - Specifies preferred tooltip placement.
* \`point\` - (optional, function) - Customizes tooltip series point rendering.
* \`header\` - (optional, function) - Renders a custom tooltip header.
* \`body\` - (optional, function) - Renders a custom tooltip body.
* \`footer\` - (optional, function) - Renders a custom tooltip footer.",
      "inlineType": {
        "name": "CartesianChartProps.TooltipOptions",
        "properties": [
          {
            "name": "body",
            "optional": true,
            "type": "((props: CartesianChartProps.TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "enabled",
            "optional": true,
            "type": "boolean",
          },
          {
            "name": "footer",
            "optional": true,
            "type": "((props: CartesianChartProps.TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "header",
            "optional": true,
            "type": "((props: CartesianChartProps.TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "placement",
            "optional": true,
            "type": ""middle" | "outside"",
          },
          {
            "name": "point",
            "optional": true,
            "type": "((props: CartesianChartProps.TooltipPointRenderProps) => BaseTooltipPointFormatted)",
          },
          {
            "name": "size",
            "optional": true,
            "type": ""small" | "medium" | "large"",
          },
        ],
        "type": "object",
      },
      "name": "tooltip",
      "optional": true,
      "type": "CartesianChartProps.TooltipOptions",
    },
    {
      "defaultValue": ""top"",
      "description": "Controls the placement of the vertical axis title.
When set to "side", displays the title along the axis line.",
      "inlineType": {
        "name": ""top" | "side"",
        "type": "union",
        "values": [
          "top",
          "side",
        ],
      },
      "name": "verticalAxisTitlePlacement",
      "optional": true,
      "type": "string",
    },
    {
      "description": "Specifies which series to show using their IDs. By default, all series are visible and managed by the component.
If a series doesn't have an ID, its name is used. When using this property, manage state updates with \`onVisibleSeriesChange\`.",
      "name": "visibleSeries",
      "optional": true,
      "type": "ReadonlyArray<string>",
    },
    {
      "description": "Defines options of the chart's x axis.
This property corresponds to [xAxis](https://api.highcharts.com/highcharts/xAxis), and extends it
with a custom value formatter.

Supported options:
* \`title\` (optional, string) - Axis title.
* \`type\` (optional, "linear" | "datetime" | "category" | "logarithmic") - Axis type.
* * "linear" - Uses continuous proportional values scale.
* * "datetime" - Similar to linear, but takes epoch time as values.
* * "category" - Uses discrete scale, requires \`categories\` to be set.
* * "logarithmic" - Uses continuous logarithmic values scale.
* \`min\` (optional, number) - Axis min value boundary.
* \`max\` (optional, number) - Axis max value boundary.
* \`tickInterval\` (optional, number) - Distance between axis ticks.
* \`categories\` (optional, Array<string>) - Predefined list of values, used for categorical axis type.
* \`valueFormatter\` (optional, function) - Takes axis tick as input and returns a formatted string. This formatter also
applies to the tooltip header.",
      "inlineType": {
        "name": "AxisOptions",
        "properties": [
          {
            "name": "categories",
            "optional": true,
            "type": "Array<string>",
          },
          {
            "name": "max",
            "optional": true,
            "type": "number",
          },
          {
            "name": "min",
            "optional": true,
            "type": "number",
          },
          {
            "name": "tickInterval",
            "optional": true,
            "type": "number",
          },
          {
            "name": "title",
            "optional": true,
            "type": "string",
          },
          {
            "name": "type",
            "optional": true,
            "type": ""category" | "datetime" | "linear" | "logarithmic"",
          },
          {
            "name": "valueFormatter",
            "optional": true,
            "type": "((value: number | null) => string)",
          },
        ],
        "type": "object",
      },
      "name": "xAxis",
      "optional": true,
      "type": "AxisOptions",
    },
    {
      "description": "Defines options of the chart's y axis.
This property corresponds to [xAxis](https://api.highcharts.com/highcharts/yAxis), and extends it
with a custom value formatter.

Supported options:
* \`title\` (optional, string) - Axis title.
* \`type\` (optional, "linear" | "datetime" | "category" | "logarithmic") - Axis type.
* * "linear" - Uses continuous proportional values scale.
* * "datetime" - Similar to linear, but takes epoch time as values.
* * "category" - Uses discrete scale, requires \`categories\` to be set.
* * "logarithmic" - Uses continuous logarithmic values scale.
* \`min\` (optional, number) - Axis min value boundary.
* \`max\` (optional, number) - Axis max value boundary.
* \`tickInterval\` (optional, number) - Distance between axis ticks.
* \`categories\` (optional, Array<string>) - Predefined list of values, used for categorical axis type.
* \`reversedStacks\` (optional, boolean) - Reverts series order in stacked series.
* \`valueFormatter\` (optional, function) - Takes axis tick as input and returns a formatted string. This formatter also
applies to the tooltip points values.",
      "inlineType": {
        "name": "CartesianChartProps.YAxisOptions",
        "properties": [
          {
            "name": "categories",
            "optional": true,
            "type": "Array<string>",
          },
          {
            "name": "max",
            "optional": true,
            "type": "number",
          },
          {
            "name": "min",
            "optional": true,
            "type": "number",
          },
          {
            "name": "reversedStacks",
            "optional": true,
            "type": "boolean",
          },
          {
            "name": "tickInterval",
            "optional": true,
            "type": "number",
          },
          {
            "name": "title",
            "optional": true,
            "type": "string",
          },
          {
            "name": "type",
            "optional": true,
            "type": ""category" | "datetime" | "linear" | "logarithmic"",
          },
          {
            "name": "valueFormatter",
            "optional": true,
            "type": "((value: number | null) => string)",
          },
        ],
        "type": "object",
      },
      "name": "yAxis",
      "optional": true,
      "type": "CartesianChartProps.YAxisOptions",
    },
  ],
  "regions": [
    {
      "description": "Custom content that renders when \`highcharts=null\`. It renders a spinner if not defined.",
      "isDefault": false,
      "name": "fallback",
    },
  ],
  "releaseStatus": "stable",
}
`;

exports[`definition for pie-chart matches the snapshot > pie-chart 1`] = `
{
  "dashCaseName": "pie-chart",
  "events": [
    {
      "cancelable": false,
      "description": "A callback function, triggered when segments visibility changes as a result of user interaction with the legend or filter.",
      "detailInlineType": {
        "name": "{ visibleSegments: Array<string>; }",
        "properties": [
          {
            "name": "visibleSegments",
            "optional": false,
            "type": "Array<string>",
          },
        ],
        "type": "object",
      },
      "detailType": "{ visibleSegments: Array<string>; }",
      "name": "onVisibleSegmentsChange",
    },
  ],
  "functions": [
    {
      "description": "Controls segments visibility and works with both controlled and uncontrolled visibility modes.",
      "name": "setVisibleSegments",
      "parameters": [
        {
          "name": "visibleSegments",
          "type": "ReadonlyArray<string>",
        },
      ],
      "returnType": "void",
    },
    {
      "description": "Functions similarly to \`setVisibleSeries\`, but applies to all series and doesn't require series IDs as input.
Use this when implementing clear-filter actions in no-match states.",
      "name": "showAllSegments",
      "parameters": [],
      "returnType": "void",
    },
  ],
  "name": "PieChart",
  "properties": [
    {
      "description": "Defines the ARIA description of the chart container.
This property corresponds to [accessibility.description](https://api.highcharts.com/highcharts/accessibility.description),
and requires the [accessibility module](https://www.highcharts.com/docs/accessibility/accessibility-module).",
      "name": "ariaDescription",
      "optional": true,
      "type": "string",
    },
    {
      "description": "Defines the ARIA label for the chart container.
This property corresponds to [lang.chartContainerLabel](https://api.highcharts.com/highcharts/lang.accessibility.chartContainerLabel),
and requires the [accessibility module](https://www.highcharts.com/docs/accessibility/accessibility-module).",
      "name": "ariaLabel",
      "optional": true,
      "type": "string",
    },
    {
      "description": "The height of the chart plot in pixels. It does not include legend and filter.",
      "name": "chartHeight",
      "optional": true,
      "type": "number",
    },
    {
      "description": "Defines the minimum allowed height of the chart plot. Use this when \`fitHeight=true\` to
prevent the chart plot from becoming too small to display its content. When the parent
container is smaller than the minimum height, a vertical scrollbar appears automatically.",
      "name": "chartMinHeight",
      "optional": true,
      "type": "number",
    },
    {
      "description": "Defines the minimum allowed width of the chart plot. When the parent container is smaller the
minimum width, the horizontal scrollbar is automatically added.",
      "name": "chartMinWidth",
      "optional": true,
      "type": "number",
    },
    {
      "description": "Defines options for filtering in the chart, including:
* \`seriesFilter\` (optional, boolean) - Displays default series filter at the top of the chart.
* \`additionalFilters\` (optional, slot) - A slot for custom chart filters at the top of the chart.",
      "inlineType": {
        "name": "BaseFilterOptions",
        "properties": [
          {
            "name": "additionalFilters",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "seriesFilter",
            "optional": true,
            "type": "boolean",
          },
        ],
        "type": "object",
      },
      "name": "filter",
      "optional": true,
      "type": "BaseFilterOptions",
    },
    {
      "defaultValue": "false",
      "description": "The chart automatically adjusts its height to fill the parent container when this property is set.",
      "name": "fitHeight",
      "optional": true,
      "type": "boolean",
    },
    {
      "description": "The Highcharts instance, which can be obtained using \`import Highcharts from "highcharts"\`.
Supported Highcharts versions: 12.",
      "name": "highcharts",
      "optional": false,
      "type": "object | null",
    },
    {
      "description": "An object that contains all of the localized strings required by the component.",
      "i18nTag": true,
      "inlineType": {
        "name": "PieI18nStrings",
        "properties": [
          {
            "name": "chartAccessibleDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "detailPopoverDismissAriaLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "errorText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "legendAriaLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "loadingText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "recoveryText",
            "optional": true,
            "type": "string",
          },
          {
            "name": "segmentAccessibleDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterLabel",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterPlaceholder",
            "optional": true,
            "type": "string",
          },
          {
            "name": "seriesFilterSelectedAriaLabel",
            "optional": true,
            "type": "string",
          },
        ],
        "type": "object",
      },
      "name": "i18nStrings",
      "optional": true,
      "type": "PieI18nStrings",
    },
    {
      "description": "Description text displayed in the donut chart's inner area.",
      "name": "innerAreaDescription",
      "optional": true,
      "type": "string",
    },
    {
      "description": "Title displayed in the donut chart's inner area.",
      "name": "innerAreaTitle",
      "optional": true,
      "type": "string",
    },
    {
      "description": "Defines chart legend options, including:
* \`enabled\` (optional, boolean) - Hides legend when set to \`false\`.
* \`title\` (optional, string) - Visible label, shown above the legend.
* \`actions\` (optional, slot) - A slot before the legend that can be used to render custom actions.",
      "inlineType": {
        "name": "BaseLegendOptions",
        "properties": [
          {
            "name": "actions",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "enabled",
            "optional": true,
            "type": "boolean",
          },
          {
            "name": "title",
            "optional": true,
            "type": "string",
          },
        ],
        "type": "object",
      },
      "name": "legend",
      "optional": true,
      "type": "BaseLegendOptions",
    },
    {
      "description": "Defines options to represent empty, no-match, loading, and error state of the chart, including:
* \`statusType\` (optional, "finished" | "loading" | "error") - Specifies the current status of loading data.
* \`empty\` (optional, ReactNode) - Content displayed when the chart data is empty.
* \`noMatch\` (optional, ReactNode) - Content displayed when there is no data to display due to the built-in filtering.
* \`loading\` (optional, ReactNode) - Content displayed when \`statusType="loading"\`. If omitted, the default loading state
is shown, using \`i18n.loadingText\` or built-in i18n.
* \`error\` (optional, ReactNode) - Content displayed when \`statusType="error"\`. If omitted, the default error state
is shown, using \`i18n.errorText\` and \`i18n.recoveryText\` (when \`onRecoveryClick\` is provided), or built-in i18n.
* \`onRecoveryClick\` (optional, function) - Called when the user clicks the recovery button that appears when using default error
state, and only if \`onRecoveryClick\` is provided. Use this to enable the user to retry a failed request or provide another option
for the user to recover from the error.",
      "inlineType": {
        "name": "BaseNoDataOptions",
        "properties": [
          {
            "name": "empty",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "error",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "loading",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "noMatch",
            "optional": true,
            "type": "React.ReactNode",
          },
          {
            "name": "onRecoveryClick",
            "optional": true,
            "type": "NonCancelableEventHandler<{}>",
          },
          {
            "name": "statusType",
            "optional": true,
            "type": ""finished" | "loading" | "error"",
          },
        ],
        "type": "object",
      },
      "name": "noData",
      "optional": true,
      "type": "BaseNoDataOptions",
    },
    {
      "description": "A function determining the segment description displayed on the chart below the segment title.",
      "inlineType": {
        "name": "(props: PieChartProps.SegmentDescriptionRenderProps) => string",
        "parameters": [
          {
            "name": "props",
            "type": "PieChartProps.SegmentDescriptionRenderProps",
          },
        ],
        "returnType": "string",
        "type": "function",
      },
      "name": "segmentDescription",
      "optional": true,
      "type": "((props: PieChartProps.SegmentDescriptionRenderProps) => string)",
    },
    {
      "description": "A function that determines the title of a segment displayed on the chart. The title appears above the segment
description defined by \`segmentDescription\`. By default, it displays the segment name. You can hide the title
by returning an empty string from the function.",
      "inlineType": {
        "name": "(props: PieChartProps.SegmentDescriptionRenderProps) => string",
        "parameters": [
          {
            "name": "props",
            "type": "PieChartProps.SegmentDescriptionRenderProps",
          },
        ],
        "returnType": "string",
        "type": "function",
      },
      "name": "segmentTitle",
      "optional": true,
      "type": "((props: PieChartProps.SegmentDescriptionRenderProps) => string)",
    },
    {
      "description": "Defines series options of the chart.
This property corresponds to [series](https://api.highcharts.com/highcharts/series), and extends it
with an additional "donut" series type.

Supported series types:
* [pie](https://api.highcharts.com/highcharts/series.pie).
* donut - the pie series with predefined inner radius.",
      "inlineType": {
        "name": "PieChartProps.SeriesOptions",
        "type": "union",
        "values": [
          "PieSeriesOptions",
          "DonutSeriesOptions",
        ],
      },
      "name": "series",
      "optional": false,
      "type": "PieChartProps.SeriesOptions",
    },
    {
      "description": "Defines tooltip options of the chart, including:
* \`enabled\` - (optional, boolean) - Hides the tooltip.
* \`size\` - (optional, "small" | "medium" | "large") - Specifies max tooltip size.
* \`details\` - (optional, function) - Provides a list of key-value pairs as tooltip's body.
* \`header\` - (optional, function) - Renders a custom tooltip header.
* \`body\` - (optional, function) - Renders a custom tooltip body.
* \`footer\` - (optional, function) - Renders a custom tooltip footer.",
      "inlineType": {
        "name": "PieChartProps.TooltipOptions",
        "properties": [
          {
            "name": "body",
            "optional": true,
            "type": "((props: TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "details",
            "optional": true,
            "type": "((props: TooltipSlotRenderProps) => ReadonlyArray<PieChartProps.TooltipDetail>)",
          },
          {
            "name": "enabled",
            "optional": true,
            "type": "boolean",
          },
          {
            "name": "footer",
            "optional": true,
            "type": "((props: TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "header",
            "optional": true,
            "type": "((props: TooltipSlotRenderProps) => React.ReactNode)",
          },
          {
            "name": "size",
            "optional": true,
            "type": ""small" | "medium" | "large"",
          },
        ],
        "type": "object",
      },
      "name": "tooltip",
      "optional": true,
      "type": "PieChartProps.TooltipOptions",
    },
    {
      "description": "Specifies which segments to show using their IDs. By default, all segments are visible and managed by the component.
If a segment doesn't have an ID, its name is used. When using this property, manage state updates with \`onVisibleSegmentsChange\`.",
      "name": "visibleSegments",
      "optional": true,
      "type": "ReadonlyArray<string>",
    },
  ],
  "regions": [
    {
      "description": "Custom content that renders when \`highcharts=null\`. It renders a spinner if not defined.",
      "isDefault": false,
      "name": "fallback",
    },
  ],
  "releaseStatus": "stable",
}
`;
